crates/proof-of-sql/src/sql/proof_exprs/test_utility.rs

```diff
@@ -218,3 +222,49 @@ pub fn sum_expr(expr: DynProofExpr, alias: &str) -> AliasedDynProofExpr {
         alias: alias.into(),
     }
 }
+
+/// Allows testing `verify_evaluate` and other verify gadgets row by row.
+/// The return matrix will tell which constraints failed.
+/// The length of the vector should be the length of the data.
+pub fn verify_row_by_row<
+    F: FnMut(&mut MockVerificationBuilder<TestScalar>, TestScalar, &[TestScalar]),
+>(
+    table_length: usize,
+    final_round_builder: &FinalRoundBuilder<'_, TestScalar>,
+    subpolynomial_max_multiplicands: usize,
+    mut row_verification: F,
+) -> Vec<Vec<bool>> {
```

Comment:
Adding support for ZeroSum would mean this return value should probably be a struct.